src/wrapper.rs

```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: MIT */
+/*
+ * Copyright (C) 2023 IBM
+ *
+ * Authors:
+ *   Claudio Carvalho <cclaudio@ibm.com>
+ *   Vikram Narayanan <vikram186@gmail.com>
+ */
+
+#![allow(non_camel_case_types)]
+
+use crate::bindings::strlen;
+use crate::mem::sizeof_alloc;
+use crate::prints;
+use crate::util::util::memset;
+use alloc::alloc::{alloc, dealloc};
+use core::alloc::Layout;
+use core::slice;
+use core::{ptr, str};
+
+pub type c_char = u8;
+
+const ALIGN_32: usize = 32;
+
+#[no_mangle]
+pub extern "C" fn malloc(sz: usize) -> *mut u8 {
+    let layout = match Layout::from_size_align(sz, ALIGN_32) {
+        Ok(l) => l,
+        Err(_e) => panic!("malloc: size is not 32 bytes aligned"),
+    };
+
+    let ptr: *mut u8 = unsafe { alloc(layout) };
+    ptr
+}
+
+#[no_mangle]
+pub extern "C" fn calloc(items: usize, sz: usize) -> *mut u8 {
+    let num_bytes: usize = items * sz;
+    let p: *mut u8 = malloc(num_bytes);
+    if !p.is_null() {
+        memset(p, 0, num_bytes);
+    }
+    p
+}
+
+#[no_mangle]
+pub extern "C" fn realloc(ptr: *mut u8, sz: usize) -> *mut u8 {
+    // If ptr is null, then the call is equivalent to malloc(sz)
+    if ptr.is_null() {
+        let new_ptr: *mut u8 = malloc(sz);
+        return new_ptr;
+    }
+
+    let ptr_size: usize = sizeof_alloc(ptr);
+    let layout = match Layout::from_size_align(ptr_size, 32) {
+        Ok(l) => l,
+        Err(_e) => panic!("realloc: ptr_size is not 32 aligned"),
+    };
+
+    // If ptr is not null and sz is zero, then the call is equivalent to free(ptr)
+    if sz == 0 {
+        unsafe {
+            dealloc(ptr, layout);
+        }
+        return ptr::null_mut();
+    }
+
+    // If the new size is larger than the old size, the added memory will not be initialized
+    // TODO: copy only the memory size requested previously rather than all the memory allocated
+    let new_ptr: *mut u8 = malloc(sz);
+    unsafe {
+        if sz < ptr_size {
+            core::ptr::copy_nonoverlapping(ptr as *const u8, new_ptr, sz)
+        } else {
+            core::ptr::copy_nonoverlapping(ptr as *const u8, new_ptr, ptr_size)
+        }
+        dealloc(ptr, layout)
+    }
+    new_ptr
+}
+
+#[no_mangle]
+pub extern "C" fn free(ptr: *mut u8) {
+    if ptr == ptr::null_mut() {
+        return;
+    }
+    let ptr_size: usize = sizeof_alloc(ptr);
```

Comment:
After the null check, you can just call mem_free() now if you use mem_allocate() (or its new forms).